src/shims/tls.rs

```diff
@@ -292,22 +324,37 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
 
     /// Schedule an active thread's TLS destructor to run on the active thread.
     /// Note that this function does not run the destructors itself, it just
-    /// schedules them one by one each time it is called.
+    /// schedules them one by one each time it is called and reenables the
+    /// thread so that it can be executed normally by the main execution loop.
     ///
     /// FIXME: we do not support yet deallocation of thread local statics.
-    fn schedule_tls_dtors_for_active_thread(&mut self) -> InterpResult<'tcx> {
+    /// Issue: https://github.com/rust-lang/miri/issues/1369
+    fn schedule_next_tls_dtor_for_active_thread(&mut self) -> InterpResult<'tcx> {
         let this = self.eval_context_mut();
         let active_thread = this.get_active_thread()?;
 
-        if this.tcx.sess.target.target.target_os == "windows" {
-            if !this.machine.tls.dtors_running.contains(&active_thread) {
-                this.machine.tls.dtors_running.insert(active_thread);
+        let scheduled_next = if this.tcx.sess.target.target.target_os == "windows" {
+            if !this.machine.tls.set_dtors_running_for_thread(active_thread) {
                 this.schedule_windows_tls_dtors()?;
+                true
+            } else {
+                false
             }
         } else {
-            this.machine.tls.dtors_running.insert(active_thread);
-            this.schedule_macos_global_tls_dtors()?;
-            this.schedule_pthread_tls_dtors()?;
+            this.machine.tls.set_dtors_running_for_thread(active_thread);
+            // The macOS thread wide destructor runs "before any TLS slots get
+            // freed", so do that first.
+            if this.schedule_macos_tls_dtor()? {
+                true
+            } else {
+                this.schedule_pthread_tls_dtors()?
+            }
+        };
```

Comment:
Why is there an asymmetry between the Windows and Unix logic here?
(Also I find this code here really hard to read, but do not have concrete proposals for improvement.)

Reply:
After thinking about it for a bit, here are some ideas (comments are just stubs):
```rust
if set_dtors_running_for_thread() {
  // First time we got here.
  if target_os == "windows" {
    // On Windows, signal that the thread quit (then come back to this scheduler).
    schedule_windows_tls_dtors();
    return;
  }
}
// See if there is a macOS dtor to run.
if this.schedule_macos_tls_dtor()? {
  // This removed the dtor -- run it, then come back (and fall through to below).
  return;
}
// See if there are still pthread dtors to run.
if this.schedule_pthread_tls_dtor()? {
  // This removed the dtor from the list -- run it, then come back.
  return;
}
// All dtors done!
this.machine.tls.delete_all_thread_tls(active_thread);
```